### ISSUE TITLE
Long loading image gallery when media is over 8K

### ISSUE THREAD

**Issue by Ava8** [DragonCherry/AssetsPickerViewController#62]
I have two devices, on each is over 8K media, so when I open assets picker, I wait for а minute or more for loading image gallery.

**Comment by Ashu** [comment]
I am having the same issue.

**Comment by DragonCherry** [comment]
@Ava8, @Ashu 
Can you please upload such images that causing the bug? I've tested with 10 images over 9MB but couldn't reproduce the problem.

**Comment by Ava8** [comment]
@DragonCherry I'm talking not about the quality but a count of photos and videos in the media library. So, my library is over 8000 photos and videos.

**Comment by DavidBoyes** [comment]
I looked in to this briefly, I currently use a different library which opens instantly where as this takes a few seconds before opening. I think it's due to the asset manager doing too much upfront. it loads every album and does an asset fetch request for each one.

**Comment by DragonCherry** [comment]
@Ava8, @Ashu , @DavidBoyes 
I couldn't produce it with iPhone 6+, over 10,000 images. "fetchAlbums" finished in very short time. Anyone can provide a fix for them?

**Comment by DavidBoyes** [comment]
I use an iPhone XS Max. ~36,000 images, iCloud enabled. Definitely noticeably slower than other libraries I have used to initially open. I'm looking in to it now. fetchAlbums takes around 0.9 seconds to complete and selectAlbum around 1.4 seconds.

**Comment by Ashu** [comment]
This issue occurred to me on a iPhone X. Currently, I don't have that device, so can't share much info about it. It works perfectly fine on my iPhone 5s with 5000 images. Also works in iPhone 6. I guess this issue occurs only on newer phones. Need to check again in new devices. Will let you know soon.

**Comment by DavidBoyes** [comment]
@Ashu Is that comparing with the same image set on all devices?

**Comment by Ashu** [comment]
> @Ashu Is that comparing with the same image set on all devices?

No. The image set was different. iPhone X was used by professional photographer, so it might have some RAW images. I need to check that.

**Comment by DavidBoyes** [comment]
> > @Ashu Is that comparing with the same image set on all devices?
> 
> No. The image set was different. iPhone X was used by professional photographer, so it might have some RAW images. I need to check that.

Ok, I tried on an iPhone 6S with ~5000 images, and it was faster but I would guess relative to the amount of images/albums rather than a device specific thing.

**Comment by rad182** [comment]
have the same issue iPhone XS having 4k++ photos. opening picker takes ~3secs.
any workaround for this?

**Comment by DavidBoyes** [comment]
A work around that removes some of the delay (not all) is to move fetchAlbums to a background thread. You'll have to manually specify the defaultAlbum in fetchAssets() so it loads the initial set of photos.

**Comment by zongziyao** [comment]
I have the same issue. 3000+ pics on my iPhone 6s and it took about 2s to open. I then wiped my iPhone and it'll open almost instantly.

**Comment by avi-screenovate** [comment]
@DragonCherry 

To get around the performance issue with loading (I have a customer with a high bar), I created a hack in my app.  I load the picker's view and then stash it as a hidden view in the main view of my app.  This causes it to load while the user doesn't notice.  Then, I add the view controller as a child view controller and move the view back.  This works surprisingly well.

My issue is that there is no way to programmatically deselect assets that are not selected in the current album.  Would it be possible to make `setSelectedAssets(assets: [PHAsset])` public, or to add a `clearAssets()` method to `AssetsPhotoViewController`?

**Comment by DragonCherry** [comment]
@avi-screenovate 
I'm sorry for this bug that make you use tricky mechanism. How about calling AssetsPhotoViewController.deselectAll()?

**Comment by avi-screenovate** [comment]
`deselectAll()` only deselects from the current Album.  It doesn't help for cross-album selections.  In my app, I now recreate the ViewController after dismissal, to hide the load time.  It's working well-enough now, and is a bit less hacky.

**Comment by eytanschulman** [comment]
Having this issue as well, takes a good 7 seconds for `AssetsPickerViewController` to load, which is quite long on my `XS`.

**Comment by gezihuzi** [comment]
Hi, I have sent an email to your mailbox (*@naver.com). Seeing this comment, can you contact me by email? I have some optimization-related questions to ask you. @DragonCherry

**Comment by DragonCherry** [comment]
Please consider replacing this library with PHPickerViewController.

**Comment by eytanschulman** [comment]
> Please consider replacing this library with PHPickerViewController.

While native solutions are always better, `PHPickerViewController` is iOS 14+, many apps will need to support iOS 13 for the time being. That's why third party solutions like this one are great.
